### PR TITLE
ingest storage: per-query `X-Read-Consistency` HTTP header

### PR DESCRIPTION
**Pull request by dimitarvdimitrov**
This PR builds on top of #7030

### What this PR does

This PR allows to specify the consistency level of individual queries. This is controlled via the HTTP header `X-Read-Consistency: strong|eventual`. Because this is still experimental, it's not documented in user-visible docs (i.e. `/docs`)

### `X-Read-Consistency` propagation

* Within a component the value is kept in the context and can be accessed via `querierapi.ReadConsistencyFromContext`
* Between Mimir components the header is propagated via an HTTP header or gRPC metadata

#### Within components

Propagating **within components** is somewhat trivial. It relies on the assumption that the original request context is retained and no place uses `context.Background()` (or an unrelated context).

#### Between components

Propagating **between components** is less trivial. Primarily because of creating phony requests, httpgrpc, and custom gRPC clients

* client --> query-frontend
	* the dskit HTTP server has a middleware which extracts the HTTP header and injects it into the HTTP request context
* query-frontend --> query-scheduler
	* query & query_range requests: the query-frontend adds the header to all httpgrpc requests
	* all other requests: the original HTTP request contains the HTTP header and that is injected in the httpgrpc request. 
	* the query-scheduler treats httpgrpc requests as opaque, so it doesn't need to understand the header
	* **Because of how all other requests are handler, the added metric `cortex_query_frontend_queries_consistency_total{consistency=""}` is only incremented for query & query_range requests, not for any other request**

* query-scheduler/query-frontend --> querier
	* after receiving the httpgrpc request from the query-frontend the querier passes it to a prometheus HTTP router, same router uses the same middleware that the query-frontend uses. This middleware parses the HTTP header and injects it into the request context
* querier --> ingester
	* the ingester gRPC client has a unary interceptor which takes the consistency level from the context and injects it as gRPC metadata
	* the dskit gRPC server has a middleware which parses the metadata and injects the consistency level in the request context

### CODE REVIEW

File: pkg/api/api.go [hunk 1]
```diff
@@ -117,6 +118,7 @@ func New(cfg Config, federationCfg tenantfederation.Config, serverCfg server.Con
 	if api.AuthMiddleware == nil {
 		api.AuthMiddleware = middleware.AuthenticateUser
 	}
+	api.AuthMiddleware = middleware.Merge(querierapi.ConsistencyMiddleware(), api.AuthMiddleware)
```
Comment (pracucci) [inline pkg/api/api.go:121]: It looks a bit weird wrapping the Auth middleware. This middleware is injected only on routes requiring authentication. Currently the consistency is just enforced on routes requiring auth, so it's not an issue, but conceptually it has nothing to do with auth (and doesn't require auth). I'm wondering if would be better just defining `ConsistencyMiddleware` on `API` and then always inject it in places where `AuthMiddleware` is injected.

**Review by pracucci — COMMENTED** [review]
I'm mid review, but I need to leave now sorry. Just committing a comment.

File: pkg/api/api.go (hunk 1 above)
Comment (dimitarvdimitrov) [reply pkg/api/api.go]: fair point. In 71ee8757dd5e760c9778ba68a1802c6603da6d86 i decided to not use indirection an directly add the `querierapi.ConsistencyMiddleware` to the handler. Let me know what you think

File: pkg/api/api.go [hunk 2]
```diff
@@ -162,6 +161,10 @@ func (a *API) RegisterRoutesWithPrefix(prefix string, handler http.Handler, auth
 func (a *API) newRoute(path string, handler http.Handler, isPrefix, auth, gzip bool, methods ...string) (route *mux.Route) {
 	if auth {
 		handler = a.AuthMiddleware.Wrap(handler)
+
+		// Assuming that we only need consistency controls when the request is for a specific tenant.
+		// Not all requests that require a tenant also support consistency, but it doesn't hurt if we propagate it anyway.
+		handler = querierapi.ConsistencyMiddleware().Wrap(handler)
```
Comment (pracucci) [inline pkg/api/api.go:167]: I think my comment is the opposite. Why not just always inject it? Also look at who else **reads** `AuthMiddleware` cause I think should inject it also when directly querying queriers for consistency.

Comment (pracucci) [reply pkg/api/api.go:167]: >Why not just always inject it?

I mean, why only when auth is used? We could just always enable it. Then the fact many APIs currently don't support it is a separate discussion, but for consistency we should always enable it.

Alternatively, we could move to the other side of the specturum and just enable it for querier's API, so only for routes registered by `RegisterQueryAPI()`.

I don't feel strong about this comment, but it still looks weird to bundle it to auth, given has nothing to do with auth.

File: pkg/frontend/querymiddleware/stats.go [hunk 3]
```diff
@@ -55,26 +63,32 @@ func (s queryStatsMiddleware) Do(ctx context.Context, req Request) (Response, er
...
-	}
-
+	s.regexpCounters(req)
+	s.consistencyCounters(ctx)
 	s.populateQueryDetails(ctx, req)
 
 	return s.next.Do(ctx, req)
 }
 
+func (s queryStatsMiddleware) regexpCounters(req Request) {
```
Comment (pracucci) [inline pkg/frontend/querymiddleware/stats.go:73]: [nit] Have you considered calling it `track....()` and `consistencyCounters` -> `track...()` too?

File: pkg/frontend/querymiddleware/stats.go [hunk 4]
```diff
@@ -107,6 +121,17 @@ func (s queryStatsMiddleware) populateQueryDetails(ctx context.Context, req Requ
 	}
 }
 
+func (s queryStatsMiddleware) consistencyCounters(ctx context.Context) {
```
Comment (pracucci) [inline pkg/frontend/querymiddleware/stats.go:124]: [nit] Would you consider improving the `Test_queryStatsMiddleware_Do`? Looks easy to do to test the new metric too.

File: pkg/ingester/active_series.go [hunk 5]
```diff
@@ -44,6 +44,12 @@ func (i *Ingester) ActiveSeries(request *client.ActiveSeriesRequest, stream clie
 		return fmt.Errorf("error parsing label matchers: %w", err)
 	}
 
+	// Enforce read consistency before getting TSDB (covers the case the tenant's data has not been ingested
+	// in this ingester yet, but there's some to ingest in the backlog).
+	if err := i.enforceReadConsistency(ctx, userID); err != nil {
```
Comment (pracucci) [inline pkg/ingester/active_series.go:49]: Thanks for adding it!

File: pkg/api/api.go (hunk 2 above)
Comment (dimitarvdimitrov) [reply pkg/api/api.go]: thanks for the review. Can you check out how dd76194ba5fa0d1d27e108d0a25f9b7dffa47f70 looks like?

File: pkg/frontend/querymiddleware/stats.go (hunk 3 above)
Comment (dimitarvdimitrov) [reply pkg/frontend/querymiddleware/stats.go]: do you mean something like `trackReadConsistency` and `trackRegexpMatchers`? 👍 done in d98612e34a75bd82127810c3601c2a8cb7e53a74

File: pkg/frontend/querymiddleware/stats.go (hunk 4 above)
Comment (dimitarvdimitrov) [reply pkg/frontend/querymiddleware/stats.go]: added in https://github.com/grafana/mimir/pull/7091/commits/bad87f1a26e3261caf5f5d6cf7518dc57f176b33

File: pkg/querier/api/consistency.go [hunk 6]
```diff
@@ -13,3 +25,84 @@ const (
...
+				r = r.WithContext(ContextWithReadConsistency(r.Context(), c))
+			}
+			next.ServeHTTP(w, r)
+		})
+	})
+}
+
+const consistencyLevelGrpcMdKey = "__consistency_level__"
+
+func ReadConsistencyClientInterceptor(ctx context.Context, method string, req any, reply any, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
```
Comment (pracucci) [inline pkg/querier/api/consistency.go:65]: [nit] Should we call it `ReadConsistencyClientUnaryInterceptor`?

File: pkg/querier/api/consistency.go [hunk 7]
```diff
@@ -13,3 +25,84 @@ const (
...
+	if c, ok := ReadConsistencyFromContext(ctx); ok {
+		ctx = metadata.AppendToOutgoingContext(ctx, consistencyLevelGrpcMdKey, c)
+	}
+	return invoker(ctx, method, req, reply, cc, opts...)
+}
+
+func ReadConsistencyServerInterceptor(ctx context.Context, req interface{}, _ *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
```
Comment (pracucci) [inline pkg/querier/api/consistency.go:72]: [nit] Should we call it `ReadConsistencyServerUnaryInterceptor`?

File: pkg/querier/worker/frontend_processor.go [hunk 8]
```diff
@@ -157,6 +157,7 @@ func (fp *frontendProcessor) runRequest(ctx context.Context, request *httpgrpc.H
 		stats, ctx = querier_stats.ContextWithEmptyStats(ctx)
 		stats.AddQueueTime(queueTime)
 	}
+	//ctx = contexWithConsistencyLevel(ctx, request.GetHeaders())
```
Comment (pracucci) [inline pkg/querier/worker/frontend_processor.go:160]: Why is it commented?

File: pkg/querier/worker/scheduler_processor.go [hunk 9]
```diff
@@ -203,6 +203,7 @@ func (sp *schedulerProcessor) querierLoop(execCtx context.Context, c schedulerpb
 
 			// We need to inject user into context for sending response back.
 			ctx = user.InjectOrgID(ctx, request.UserID)
+			//ctx = contexWithConsistencyLevel(ctx, request.GetHttpRequest().GetHeaders())
```
Comment (pracucci) [inline pkg/querier/worker/scheduler_processor.go:206]: Why is it commented?

**Review by pracucci — APPROVED** [review]
I just left a comment about a couple of commented lines of code. We don't have any e2e test (which I believe would be the only way to test everything is wiring correctly), but I don't think that's a blocker. We can move forward and we can add "adding e2e tests" to an issue for the future.

File: pkg/querier/worker/frontend_processor.go (hunk 8 above)
Comment (dimitarvdimitrov) [reply pkg/querier/worker/frontend_processor.go]: 😬 this PR was a mess, sorry about that

**Pull request merged**